---
title: "CoGAPS - Coordinated Gene Association in Pattern Sets"
author: "Jeanette Johnson, Ashley Tsang, Thomas Sherman, Genevieve Stein-O'Brien, Hyejune Limb, Elana Fertig"
date: "`r BiocStyle::doc_date()`"
bibliography: References.bib
vignette: >
    %\VignetteIndexEntry{CoGAPS}
    %\VignetteEncoding{UTF-8}
    %\VignetteEngine{knitr::rmarkdown}
output: 
    BiocStyle::html_document
editor_options: 
  markdown: 
    wrap: 72
---

```{r include=FALSE, cache=FALSE}
library(CoGAPS)
library(BiocParallel)
```

# Vignette Version

This vignette was built using CoGAPS version:

```{r}
packageVersion("CoGAPS")
```

# Introduction

Coordinated Gene Association in Pattern Sets (CoGAPS) is a technique for
latent space learning in gene expression data. CoGAPS is a member of the
Nonnegative Matrix Factorization (NMF) class of algorithms. NMFs
factorize a data matrix into two related matrices containing gene
weights, the Amplitude (A) matrix, and sample weights, the Pattern (P)
Matrix. Each column of A or row of P defines a feature and together this
set of features defines the latent space among genes and samples,
respectively. In NMF, the values of the elements in the A and P matrices
are constrained to be greater than or equal to zero. This constraint
simultaneously reflects the non-negative nature of gene expression data
and enforces the additive nature of the resulting feature dimensions,
generating solutions that are biologically intuitive to interpret
(@SEUNG_1999).

CoGAPS has two extensions that allow it to scale up to large data sets,
Genome-Wide CoGAPS (GWCoGAPS) and Single-Cell CoGAPS (scCOGAPS). This
package presents a unified R interface for all three methods, with a
parallel, efficient underlying implementation in C++.

# Installing CoGAPS

*CoGAPS* is a bioconductor package and so the release version can be
installed as follows:

```{r eval=FALSE}
# if (!require("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# 
# BiocManager::install("CoGAPS")

## Method 2 using devtools package
devtools::install_github("FertigLab/CoGAPS")
```

# Package Overview

We first give a walkthrough of the package features using a simple,
simulated data set. In later sections we provide two example workflows
on real data sets.

## Running CoGAPS with Default Parameters

The only required argument to `CoGAPS` is the data set. This can be a
`matrix`, `data.frame`, `SummarizedExperiment`, `SingleCellExperiment`
or the path of a file (`tsv`, `csv`, `mtx`, `gct`) containing the data.

```{r}
library(CoGAPS)
```

```{r modsim}
modsimdata <- read.table("../data/ModSimData.txt")
# input to CoGAPS
modsimdata

# create new parameters object
params <- new("CogapsParams")

# view all parameters
params

# get the value for a specific parameter
getParam(params, "nPatterns")

# set the value for a specific parameter
params <- setParam(params, "nPatterns", 3)
getParam(params, "nPatterns")
```

Once we've created the parameters object we can pass it along with our
data to `CoGAPS`.

```{r}
# run CoGAPS with specified parameters
cogapsresult <- CoGAPS(modsimdata, params, outputFrequency = 10000)
cogapsresult
```

There are several other arguments that are passed directly to `CoGAPS`
which are covered in later sections.

## Breaking Down the Return Object from CoGAPS

CoGAPS returns a object of the class `CogapsResult` which inherits from
`LinearEmbeddingMatrix` (defined in the `SingleCellExperiment` package).
CoGAPS stores the lower dimensional representation of the samples (P
matrix) in the `sampleFactors` slot and the weight of the features (A
matrix) in the `featureLoadings` slot. `CogapsResult` also adds two of
its own slots - `factorStdDev` and `loadingStdDev` which contain the
standard deviation across sample points for each matrix.

There is also some information in the `metadata` slot such as the
original parameters and value for the Chi-Sq statistic. In general, the
metadata will vary depending on how `CoGAPS` was called in the first
place. The package provides these functions for querying the metadata in
a safe manner:

```{r single cell}
options(timeout=1000) # adjust this if you're getting timeout downloading the file
url = "https://zenodo.org/record/7709664/files/inputdata.Rds?download=1"
download.file(url,"inputdata_download.Rds")
pdac_data <- readRDS("inputdata_download.Rds")

pdac_epi_counts <- counts(pdac_data)
pdac_epi_counts <- as.matrix(pdac_data@assays$originalexp@counts)
head(pdac_epi_counts, n=c(5L, 2L))

library(CoGAPS)
pdac_params <- CogapsParams(nIterations=100, # run for 100 iterations 
                            seed=42, # set random seed for consistent results across stochastic runs
                            nPatterns=8, # each thread will learn 8 patterns
                            sparseOptimization=TRUE, # optimize for sparse input data
                            distributed="genome-wide") # paralellize across sets "genome-wise
pdac_params
pdac_params <- setDistributedParams(pdac_params, nSets=7)
pdac_params
startTime <- Sys.time()

vignette_result <- CoGAPS(pdac_epi_counts, pdac_params)
endTime <- Sys.time()
  
# prints recorded time
print(endTime - startTime)
```

# sessionInfo()

```{r}
library(Matrix)
library(CoGAPS)
options(timeout=1000) # adjust this if you're getting timeout downloading the file
url = "https://zenodo.org/record/7709664/files/cogapsresult.Rds?download=1"
download.file(url,"cogapsresult_download.Rds")
cogapsresult <- readRDS("cogapsresult_download.Rds")

plotPatternUMAP(vignette_result, inputdata)

library(gridExtra)
library(ggplot2)
# Merge with the column data in the cds
factor_columns <- colnames(cogapsresult@sampleFactors)
pattern_names <- factor_columns[grep("Pattern", factor_columns)]
inputdata@colData <- cbind(inputdata@colData, cogapsresult@sampleFactors[,pattern_names ])

plot_cells(inputdata, 
           color_cells_by = "TN_assigned_cell_type_immune_broad",
           label_cell_groups = FALSE, show_trajectory_graph = FALSE) +
           ggtitle(paste0(pattern)) +
           theme(legend.position="none")


umaplist <- list()
for (pattern in pattern_names)
  # UMAP Embedding of CoGAPS Pattern Weight ###################################
  umaplist[[length(umaplist)+1]] <- plot_cells(inputdata, color_cells_by = pattern,
                      label_cell_groups = FALSE, show_trajectory_graph = FALSE) +
                      ggtitle(paste0(pattern)) +
                      theme(legend.position="none")

plot <- grid.arrange(grobs = umaplist)
plot

cds2 <- subset(cds, gene_short_name %in% rownames(output@featureLoadings))
countsdf=as.data.frame(as.matrix(cds2@assays@data$counts))
write.csv(countsdf, "/Users/jeanette/databarn/pdac_atlas_cogaps/aws_result/counts.csv")
```
```{r trying seurat input object and analysis}
library(Seurat)
inputdata <- readRDS("../data/inputdata.Rds")
pdac_data <- readRDS("../data/inputdata.Rds")
inputdata
inputdata <- NormalizeData(inputdata)
inputdata <- FindVariableFeatures(inputdata)
inputdata <- ScaleData(inputdata)
inputdata <- RunPCA(inputdata)
inputdata <- RunUMAP(inputdata, dims=1:15)

UMAPPlot(inputdata)
patterns_in_order <- t(cogapsresult@sampleFactors[colnames(inputdata),])
# add CoGAPS patterns as an assay
inputdata[["CoGAPS"]] <- CreateAssayObject(counts = patterns_in_order)
DefaultAssay(inputdata) <- "CoGAPS"
pattern_names = rownames(inputdata@assays$CoGAPS)
library(viridis)
color_palette <- viridis(n=10)
FeaturePlot(inputdata, pattern_names, cols=color_palette, reduction = "umap") & NoLegend() 

saveRDS(inputdata, "../data/inputdata_seurat.Rds")
```
# Citing CoGAPS

If you use the CoGAPS package for your analysis, please cite
@FERTIG_2010

If you use the gene set statistic, please cite @OCHS_2009

# References

