---
title: "CoGAPS Workflow Vignette"
author: "Hyejune Limb"
date: "`r doc_date()`"
package: "`r pkg_ver('CoGAPS')`"
bibliography: AppNote.bib
vignette: >
    %\VignetteIndexEntry{GWCoGAPS and PatternMarkers}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
output: 
    BiocStyle::html_document
---

```{r include=FALSE, cache=FALSE}
library(CoGAPS)
```

# Introduction

Gene Association in Pattern Sets (GAPS) infers underlying patterns in a matrix of measurements that can
be interpreted as arising from the multiplication of two lower dimensional matrices. The first development of
this code in R/Biocondcutor was focused on gene expression analysis, however the original concept was used
in spectral imaging. The approach is a general form of matrix factorization using a stochastic algorithm. In
this vignette, we will focus on gene expression analysis for concreteness, but the factorization is applicable
more broadly.

  The Markov chain Monte Carlo (MCMC) matrix factorization that infers patterns also infers the extent
to which individual genes belong to these patterns. The CoGAPS algorithm extends GAPS to infer the
coordinated activity in sets of genes for each of the inferred patterns based upon (5) and to refine gene set
membership based upon (2).

  The GAPS algorithm is implemented in C++ and compiled and integrated into R using the Rcpp package.
GAPS is licensed under the GNU General Public License version 2. You may freely modify and redistribute
GAPS under certain conditions that are described in the top level source directory file COPYING.

  The R package CoGAPS is designed to facilitate the corresponding analysis of microarray measurements
by calling the GAPS C++ library. With the migration to C++ code, the installation as noted in Chapter 2
should now be automatic. Running instructions for the GAPS and CoGAPS analyses are provided in Sections
3.1 and 3.2, respectively. CoGAPS and GAPS are freely available at https://github.com/ejfertig/CoGAPS
and through the CoGAPS Bioconductor package.

  If you use the CoGAPS package for your analysis please cite: (1) EJ Fertig, J Ding, AV Favorov, G
Parmigiani, and MF Ochs (2010) CoGAPS: an R/C++ package to identify patterns and biological process
activity in transcriptomic data. Bioinformatics 26: 2792-2793.

  To cite the CoGAPS algorithm use: (3) MF Ochs (2003) Bayesian Decomposition in The Analysis of
Gene Expression Data: Methods and Software G Parmigiani, E Garrett, R Irizarry, and S Zeger, ed. New
York: Springer Verlag.

  To cite the gene set statistic use: (5) MF Ochs, L Rink, C Tarn, S Mburu, T Taguchi, B Eisenberg, and
AK Godwin (2009) Detection of treatment-induced changes in signaling pathways in gastrointestinal stromal
tumors using transcriptomic data. Cancer Research 69: 9125-9132.

  To site the set-membership refinement statistic use: (2) EJ Fertig, AV Favorov, and MF Ochs (2012)
Identifying context-specific transcription factor targets from prior knowledge and gene expression data. 2012
IEEE International Conference on Bioinformatics and Biomedicine, B310, in press.
Please contact Elana J. Fertig ejfertig@jhmi.edu or Michael F. Ochs ochsm@tcnj.edu for assistance.

# CoGAPS
Here we read data into an R object, but we are currently adding support for passing a file name (.csv, .tsv, .mtx) for the data. This data comes with an uncertainty matrix (GIST.S) but we are also working on allowing just the data matrix to be passed.
```{r}
suppressMessages(library(CoGAPS))
print(packageVersion("CoGAPS"))
```

```{r}
data("GIST")
```

## Running CoGAPS
Here we run CoGAPS on the data set `GIST.D` for a few different values of nFactor (if the data matrix is M x N, then we find D = A * P, where A is M x nFactor and P is nFactor x N). Typically we run for 50000 iterations but we are also working on dynamically selecting the number of iterations so that the user can just pass a “max iterations” parameter if they choose, otherwise it will be handled internally.

```{r}
nIterations <- 2500
patternRange <- 2:4
cogapsResult <- list()
for (nPatterns in patternRange)
{
    cogapsResult[[nPatterns]] <- CoGAPS(as.matrix(GIST.D),
        as.matrix(GIST.S), nFactor=nPatterns, nSample=nIterations,
        nEquil=nIterations, seed=123, messages=FALSE)
}
```

## Selecting Number of Patterns
Selecting the best value for nFactor (i.e. number of patterns) is the most difficult part of the analysis. The user may have different ways of selecting dimensionality. We are working on a more formal process for this. Here we show the simplest approach of plotting the error and selecting the least number of patterns that sufficiently reduce the error.

```{r}
chisq <- sapply(patternRange, function(i) cogapsResult[[i]]$meanChi2)
plot(patternRange, chisq)
```

```{r}
bestResult <- cogapsResult[[3]]
```

## Plotting outputs
```{r}
# pdf("GIST_example.pdf") # uncomment to save file to pdf
plotP(bestResult$Pmean)
# dev.off() # uncomment to save file to pdf
```


# Using CoGAPS-based statistics to infer gene membership in annotated gene sets

## CoGAPSStat
The function `calcCoGAPSStat` is used to infer gene set activity in each pattern from the GAPS matrix factorization.
$$ calcCoGAPSStat(Amean, Asd, GStoGenes, numPerm=500) $$
`calcCoGAPSStat` calculates the gene set statistics for each column of a matrix A using a Z-score from the elements of the A matrix, the input gene set and permutation tests. The function outputs a list containing GSUpreg, GSDownreg, and GSActEst. 

* *GSUpreg* lists p-values for upregulation of each gene set in each pattern

* *GSDownreg* lists p-values for downregulation of each gene set in each pattern

* *GSActEst* provides gene set activity through conversion of p-values to activity estimates of each gene set in each pattern

```{r}
data("SimpSim")
calcCoGAPSStat(SimpSim.result$Amean, SimpSim.result$Asd, GStoGenes=GSets, numPerm=500)
```

## computeGeneGSProb
Now using the `computeGeneGSProb` function, we can use the gene set statistic (returned from calcCoGAPSStat) to compute a statistic to quantify the likely membership of each gene annotated to a set based on its inferred activity. 
$$ computeGeneGSProb(Amean, Asd, GStoGenes, numPerm=500) $$
The statistic used to infer membership compares the expression pattern of a gene annotated as a member of a gene set to the common expression pattern of all annotated members of that gene set. The function outputs the p-value of a set membership for each gene specified in GStoGenes. 
